src/librustc/infer/fudge.rs

```diff
@@ -124,6 +121,40 @@ impl<'a, 'gcx, 'tcx> TypeFolder<'gcx, 'tcx> for RegionFudger<'a, 'gcx, 'tcx> {
         self.infcx.tcx
     }
 
+    fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+        match ty.sty {
+            ty::TyInfer(ty::InferTy::TyVar(vid)) => {
+                match self.type_variables.get(&vid) {
```

Comment:
this function is already a total hack, but I'm afraid this makes behavior depend on the mood of the union-find algorithm: if a set of (possibly-preexisting) type variables is equated and returned, we'll return one of them "at random". 

For example, in
```Rust
use std::mem;

fn main() {
    let v = Vec::new();
    fn example<'a, U>(x: Option<&'a mut &[u8]>, u: U) -> &'a mut U { loop {} }
    let mut x: Option<&mut $0> = None;
    let mut y = None;
    mem::swap(&mut x, &mut y); // equate types, so `y: Option<&mut $0>`.
    *x.unwrap() = *example::<$1>(y, &v);
}
```

When this is checked, EIfEO unifies `$0` and `$1` within the snapshot (a unification that is then discarded), and checks `u` with either of the type variables as the expected var, depending on what union-find picks as the root. `$0` is unified later on with `&[u8]`, so if `$0` is picked then the coercion from `&Vec<u8>` to `&[u8]` occurs before the call to `example`, and if `$1` is picked the coercion occurs after the call.

This might lead to instability when union-find picks different roots. Not sure how much important is this.